runtime/rust_backend/Runtime.cpp

```diff
@@ -0,0 +1,357 @@
+// This file is part of SymCC.
+//
+// SymCC is free software: you can redistribute it and/or modify it under the
+// terms of the GNU General Public License as published by the Free Software
+// Foundation, either version 3 of the License, or (at your option) any later
+// version.
+//
+// SymCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// SymCC. If not, see <https://www.gnu.org/licenses/>.
+
+#include <Runtime.h>
+#include <RustRuntime.h>
+
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <set>
+#include <vector>
+
+#ifndef NDEBUG
+#include <chrono>
+#endif
+
+#include "Config.h"
+#include "GarbageCollection.h"
+#include "LibcWrappers.h"
+#include "Shadow.h"
+
+#ifndef NDEBUG
+// Helper to print pointers properly.
+#define P(ptr) reinterpret_cast<void *>(ptr)
+#endif
+
+/* TODO Eventually we'll want to inline as much of this as possible. I'm keeping
+   it in C for now because that makes it easier to experiment with new features,
+   but I expect that a lot of the functions will stay so simple that we can
+   generate the corresponding bitcode directly in the compiler pass. */
+
+namespace {
+
+/// Indicate whether the runtime has been initialized.
+std::atomic_flag g_initialized = ATOMIC_FLAG_INIT;
+
+FILE *g_log = stderr;
+
+#ifndef NDEBUG
+[[maybe_unused]] void dump_known_regions() {
+  std::cerr << "Known regions:" << std::endl;
+  for (const auto &[page, shadow] : g_shadow_pages) {
+    std::cerr << "  " << P(page) << " shadowed by " << P(shadow) << std::endl;
+  }
+}
+
+#endif
+
+/// The set of all expressions we have ever passed to client code.
+std::set<SymExpr> allocatedExpressions;
```

Comment:
It would be very nice if we could refactor the parts that are the same in the simple backend and your new backend into a shared header or something of the sort, in order to avoid code duplication. We could even go as far as expressing the simple backend as an implementation of your interface - which would nicely solve @aurelf's demand for a testable version of the new backend.